immer/detail/hamts/bits.hpp

```diff
@@ -18,43 +18,91 @@ namespace immer {
 namespace detail {
 namespace hamts {
 
-using bits_t   = std::uint32_t;
-using bitmap_t = std::uint32_t;
-using count_t  = std::uint32_t;
-using shift_t  = std::uint32_t;
 using size_t   = std::size_t;
 using hash_t   = std::size_t;
 
-template <bits_t B, typename T=count_t>
+template <size_t B>
+struct types
+{
+    static_assert(B < 6, "B > 6 is not supported.");
+
+    using bits_t   = std::uint32_t;
+    using bitmap_t = std::uint32_t;
+    using count_t  = std::uint32_t;
+    using shift_t  = std::uint32_t;
+};
```

Comment:
Hi @fedormatantsev!
One thing that is important: only `bitmap_t` depends on `B`! The other types should remain outside of this `types` template. In this way, you can also keep the `bits_t` type, which you have removed. Also in that way you need to change way less things overall in the code. There are very few mentions to `bitmap_t` in the code, which is the only type dependent on `B`.

Reply:
Yeah, sounds reasonable.